packages/gatsby/src/commands/serve.js

```diff
@@ -74,27 +90,29 @@ module.exports = async program => {
   })
   app.use(pathPrefix, router)
 
-  const server = app.listen(port, host, () => {
-    let openUrlString = `http://${host}:${port}${pathPrefix}`
-    console.log(
-      `${chalk.blue(`info`)} gatsby serve running at: ${chalk.bold(
-        openUrlString
-      )}`
-    )
-    if (open) {
-      console.log(`${chalk.blue(`info`)} Opening browser...`)
-      Promise.resolve(openurl(openUrlString)).catch(err =>
-        console.log(
-          `${chalk.yellow(
-            `warn`
-          )} Browser not opened because no browser was found`
-        )
+  const startListening = () => {
+    app.listen(port, host, () => {
+      let openUrlString = `http://${host}:${port}${pathPrefix}`
+      console.log(
+        `${chalk.blue(`info`)} gatsby serve running at: ${chalk.bold(
+          openUrlString
+        )}`
       )
-    }
-  })
+      if (open) {
+        console.log(`${chalk.blue(`info`)} Opening browser...`)
+        Promise.resolve(openurl(openUrlString)).catch(err =>
+          console.log(
+            `${chalk.yellow(
+              `warn`
+            )} Browser not opened because no browser was found`
+          )
+        )
+      }
+    })
+  }
 
-  signalExit(() => {
-    telemetry.trackCli(`SERVE_STOP`)
-    server.close()
+  detectPortInUseAndPrompt(port, rlInterface, newPort => {
```

Comment:
if we go for a promise interface we can write it as
`port = await detectPortInUseAndPrompt(port, rlInterface, newPort)`